src/glang/codegen/mangling.py

```diff
@@ -0,0 +1,189 @@
+from typing import Iterable
+
+from .interfaces import SpecializationItem, Type
+
+
+def mangle_int(size_in_bits: int, is_signed: bool) -> str:
+    # <builtin-type> ::= c  # char
+    #                ::= a  # signed char
+    #                ::= h  # unsigned char
+    #                ::= s  # short
+    #                ::= t  # unsigned short
+    #                ::= i  # int
+    #                ::= j  # unsigned int
+    #                ::= l  # long
+    #                ::= m  # unsigned long
+    #                ::= x  # long long, __int64
+    #                ::= y  # unsigned long long, __int64
+    #                ::= n  # __int128
+    #                ::= o  # unsigned __int128
+    match size_in_bits, is_signed:
+        case 8, True:
+            return "c"
+        case 8, False:
+            return "h"
+        case 16, True:
+            return "s"
+        case 16, False:
+            return "t"
+        case 32, True:
+            return "i"
+        case 32, False:
+            return "j"
+        case 64, True:
+            return "l"
+        case 64, False:
+            return "m"
+        case 128, True:
+            return "n"
+        case 128, False:
+            return "o"
+
+    raise AssertionError()
+
+
+def mangle_float(size_in_bits: int) -> str:
+    # <builtin-type> ::= DF <number> _  # _FloatN (N bits)
+    #                ::= f              # float
+    #                ::= d              # double
+    #                ::= g              # __float128
+    match size_in_bits:
+        case 16:
+            return "DF16_"
+        case 32:
+            return "f"
+        case 64:
+            return "d"
+        case 128:
+            return "g"
+
+    raise AssertionError()
+
+
+def mangle_source_name(identifier: str) -> str:
+    assert len(identifier) > 0
+
+    # <source-name> ::= <positive length number> <identifier>
+    return f"{len(identifier)}{identifier}"
+
+
+def mangle_template_arg(arg: SpecializationItem) -> str:
+    if isinstance(arg, Type):
+        # <template-arg> ::= <type>  # type or template
+        return arg.ir_mangle
+    else:
+        assert isinstance(arg, int)
+        # <template-arg> ::= <expr-primary>             # simple expressions
+        # <expr-primary> ::= L <type> <value number> E  # integer literal
+        # FIXME mangle as an int, even though we don't actually force a type.
+        # TODO negative numbers.
+        return f"Li{arg}E"
```

Comment:
Wow, this spec is involved. Looks like negative numbers are quite easy to support tho.

Reply:
Yep, they thought of everything :)

I'll get around to adding negative numbers soon, I just didn't want to do it in this PR